### PR TITLE
23/copy

### LINKED ISSUE #22: Handle styles correctly with SSR
This requires to apply SSR correctly for Material styles  https://material-ui.com/styles/advanced/#next-js

### LINKED ISSUE #23: Copy Item in Graasp

### PR DESCRIPTION
**Pull request by pyphilia**
close #23
close #22 

contains #21

### CODE REVIEW

File: src/client/components/collection/CopyButton.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  }, [isSignedIn]);
+
+  const onClick = async () => {
+    if (!isSignedIn) {
+      setShowLoginModal(true);
+    } else {
+      setShowTreeModal(true);
+    }
+
+    setShowTreeModal(true);
```
Comment (pyphilia) [inline src/client/components/collection/CopyButton.js:66]: remove this line.

**Comment by pyphilia** [comment]
This PR is ready for review, but you can review https://github.com/graasp/graasp-explore/pull/21/files to review a first part, then this PR will be rebased.

It contains the possibility to copy a space in your own workspace, and provides minimal feedback on success or fail (eg. ILS cannot be copied in another ILS).

File: src/api/collection.js [hunk 2]
```diff
@@ -1,21 +1,48 @@
...
+    body: JSON.stringify(body),
+    headers: {
+      ...DEFAULT_POST.headers,
+      cookie,
+    },
+  });
+
+  // error if the !res.ok
+  // or redirect the request to the login page
+  if (!res.ok || res.url?.includes('/login')) {
```
Comment (juancarlosfarah) [inline src/api/collection.js:42]: Factor out `/login`.

File: src/client/components/Root.js [hunk 3]
```diff
@@ -52,20 +52,27 @@ let theme = createMuiTheme({
...
-    <MuiThemeProvider theme={theme}>
-      <I18nextProvider i18n={i18nConfig}>
-        <App />
-        <ToastContainer />
-      </I18nextProvider>
-    </MuiThemeProvider>
-  </div>
-);
+const Root = ({ classes }) => {
+  useEffect(() => {
```
Comment (juancarlosfarah) [inline src/client/components/Root.js:56]: Can you comment why we need this effect?

File: src/client/components/collection/CollectionsGrid.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import { useTranslation } from 'react-i18next';
+import Typography from '@material-ui/core/Typography';
+import Grid from '@material-ui/core/Grid';
+import CollectionCard from './CollectionCard';
+
+function CollectionsGrid({ collections }) {
+  const { t } = useTranslation();
+
+  return !collections.length ? (
+    <div className="Main">
```
Comment (juancarlosfarah) [inline src/client/components/collection/CollectionsGrid.js:12]: What's this class name?

File: src/client/components/collection/CopyButton.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  const [showTreeModal, setShowTreeModal] = useState(false);
+  const [isSignedIn, setIsSignedIn] = useState(false);
+  const [copying, setCopying] = useState(false);
+
+  const checkUserIsSignedIn = async () => {
+    const res = await (await fetch(IS_AUTHENTICATED_ROUTE, DEFAULT_GET)).json();
+    setIsSignedIn(res);
+  };
+
+  useEffect(async () => {
```
Comment (juancarlosfarah) [inline src/client/components/collection/CopyButton.js:46]: I really like this!

File: src/client/components/collection/CopyButton.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    if (!isSignedIn) {
+      setShowLoginModal(true);
+    } else {
+      setShowTreeModal(true);
+    }
+
+    setShowTreeModal(true);
+  };
+
+  const copy = async ({ id: toSpace }) => {
```
Comment (juancarlosfarah) [inline src/client/components/collection/CopyButton.js:69]: Do we not want to factor out these calls that go to the backend ?

File: src/server.js [hunk 7]
```diff
@@ -64,34 +89,89 @@ const handleRender = (req, res, data) => {
   }
 };
 
+const handleSpaceVieverRender = (req, res) => {
```
Comment (juancarlosfarah) [inline src/server.js:92]: typo in `Viewer`

File: src/index.js [hunk 8]
```diff
@@ -4,6 +4,7 @@ let app = require('./server').default;
 
 if (module.hot) {
   module.hot.accept('./server', () => {
+    // eslint-disable-next-line no-console
```
Comment (juancarlosfarah) [inline src/index.js:7]: Add an issue to add a logger like the one we have for the API. Then we won't get any of these errors.

File: src/client/utils/common.js [hunk 9]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line import/prefer-default-export
+export const cloneDeep = (el) => JSON.parse(JSON.stringify(el));
```
Comment (juancarlosfarah) [inline src/client/utils/common.js:2]: Can't use anything in `lodash`? `JSON.parse` and `JSON.stringify` can throw errors.

File: src/client/config/helpers.js [hunk 10]
```diff
@@ -1,11 +1,67 @@
...
 
-export {
-  // eslint-disable-next-line import/prefer-default-export
-  openInNewTab,
+const openContentInNewTab = (content) => {
+  const newWindow = window.open('', '_blank');
+  newWindow.document.write(content);
 };
+
+const openInNewWindow = ({ url, title }) => {
```
Comment (juancarlosfarah) [inline src/client/config/helpers.js:15]: Is this still being used?

File: src/client/config/env.js [hunk 11]
```diff
@@ -1,2 +1,5 @@
-// eslint-disable-next-line import/prefer-default-export
-export const { REACT_APP_GOOGLE_ANALYTICS_ID } = process.env;
+// export const {
```
Comment (juancarlosfarah) [inline src/client/config/env.js:1]: Add an issue to solve adding env files into the front end. Probably requires some Webpack stuff. @hasanagh, any ideas?

File: src/client/config/constants.js [hunk 12]
```diff
@@ -28,3 +28,13 @@ export const MEMBER_TYPES = {
...
   CONTRIBUTOR: 'contributor',
 };
+
+export const LOGIN_POP_UP_WIDTH = 450;
+export const LOGIN_POP_UP_HEIGHT = 450;
+
+export const ROOT_ID = 'ROOT';
+export const TREE_VIEW_HEIGHT = 300;
+export const TREE_VIEW_MIN_WIDTH = 350;
+
```
Comment (juancarlosfarah) [inline src/client/config/constants.js:38]: For the following you can have multiple errors and success codes. Try to use ranges or `response.ok`.

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
Very nice! I love this PR! :-)

Just some minor questions and comments.

As an aside, try to add comments in the tricky bits. To feature-proof.

File: Dockerfile [hunk 13]
```diff
@@ -1,16 +1,20 @@
...
+RUN mkdir -p /usr/src/app
+WORKDIR /usr/src/app
+
+# install app dependencies
+COPY package.json /usr/src/app/
+RUN yarn install
+
+# bundle app source
+COPY . /usr/src/app
+
```
Comment (hasanagh) [inline Dockerfile:14]: I dont think we need to specify the full absolute path after defining the Workdir step. you can just use `.`

**Review by hasanagh — CHANGES_REQUESTED** [review]
Thanks for the PR! I had one comment in addition to JC's

File: src/client/components/Root.js (hunk 3 above)
Comment (hasanagh) [reply src/client/components/Root.js:56]: Should this effect run on each render? cant have an empty dependency list?

File: src/client/components/collection/CopyButton.js (hunk 5 above)
Comment (hasanagh) [reply src/client/components/collection/CopyButton.js:46]: I think and in order to use an async call inside you effect. you should be doing something like this:
`useEffect(() => {`
  `  (async function checkSignIn() {`
      `await checkUserIsSignedIn();`
  `  })();`
`  }, []);`

check this: https://github.com/facebook/react/issues/14326#issuecomment-441680293

File: src/client/config/constants.js (hunk 12 above)
Comment (pyphilia) [reply src/client/config/constants.js:38]: These are used to set express response, I don't think I can only set `ok` but if there is a way let me know! I will rename `DEFAULT_SUCCESS_CODE` and forward the status code from the original call.

**Comment by pyphilia** [comment]
I deployed the latest version and apply all the requested changes. I also edited:
- internal requests response. I return a JSON object which is more clean.
- the tree modal switches to login modal when the user get signed out


Side note: the copy actually copy ILS in ILS successfully, but André is working on this.

File: src/api/authentication.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+import fetch from 'node-fetch';
+import { DEFAULT_ERROR_CODE } from '../client/config/constants';
+import { DEFAULT_GET, formatCookies } from './common';
+import { IS_AUTHENTICATED_ENDPOINT } from './endpoints';
+
+// eslint-disable-next-line import/prefer-default-export
+export const isAuthenticated = async (cookies) => {
```
Comment (juancarlosfarah) [inline src/api/authentication.js:7]: Maybe a short comment explaining what this is doing.

File: src/api/navigation.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
+import fetch from 'node-fetch';
+import { DEFAULT_ERROR_CODE } from '../client/config/constants';
+import { DEFAULT_GET, formatCookies } from './common';
+import { GET_NAV_TREE_ENDPOINT } from './endpoints';
+
+// eslint-disable-next-line import/prefer-default-export
+export const getNavTree = async (cookies) => {
```
Comment (juancarlosfarah) [inline src/api/navigation.js:7]: Maybe just a quick comment saying what this is fetching.

File: src/server.js [hunk 16]
```diff
@@ -65,20 +88,57 @@ const handleRender = (req, res, data) => {
   }
 };
 
+// redirect to viewer mode of the space
+const handleSpaceViewerRender = (req, res) => {
+  const { id } = req.params;
+  if (!ObjectId.isValid(id)) {
```
Comment (juancarlosfarah) [inline src/server.js:94]: If you're throwing an error here, you should catch it with some middleware so that the server doesn't collapse.

Comment (juancarlosfarah) [reply src/server.js:94]: Are you doing that?

File: src/client/config/constants.js (hunk 12 above)
Comment (juancarlosfarah) [reply src/client/config/constants.js:35]: I see. OK. In that case, you should be explicit. I would use something like this:

https://www.npmjs.com/package/http-status-codes

**Review by juancarlosfarah — CHANGES_REQUESTED** [review]
Looks good! Happy you got it to work. Just some minor questions, comments and suggestions here.

File: src/server.js (hunk 16 above)
Comment (pyphilia) [reply src/server.js:94]: No, it just fails. Maybe we can redirect the user to a 404 page. I'll open an issue with this.

**Pull request merged**